### PR TITLE
Revise reference implementation for ReduceMax operation

### PR DESCRIPTION
**Pull request by ggalieroc**
### Details: Reference Implementation of ReduceMax refactored.
 - Remove CoordinateTransform calls to index(). Benchmark results in 45971.
 - Remove keep_dims from ref impl signature.
 - Moved backend tests with zero dims and dynamic shape to op_eval.
 - Removed backend tests with double precision.
 
### Tickets:
 - *37467*, *45971*

### CODE REVIEW

File: ngraph/core/reference/include/ngraph/runtime/reference/max.hpp [hunk 1]
```diff
@@ -17,35 +18,33 @@ namespace ngraph
...
+                std::fill(out, out + shape_size(out_shape), minval);
 
-                for (const Coordinate& output_coord : output_transform)
-                {
-                    out[output_transform.index(output_coord)] = minval;
-                }
+                const auto in_strides = row_major_strides(in_shape);
+                const auto out_strides = row_major_strides(out_shape);
 
                 CoordinateTransform input_transform(in_shape);
```
Comment (ilyachur) [inline ngraph/core/reference/include/ngraph/runtime/reference/max.hpp:33]: Can we avoid `CoordinateTransform`?

Comment (ggalieroc) [reply ngraph/core/reference/include/ngraph/runtime/reference/max.hpp]: We discussed this in different op, ReduceSum. The conclusion was that we cannot avoid it fully. Here the [comment](https://github.com/openvinotoolkit/openvino/pull/5728#discussion_r639599572)

File: ngraph/core/src/op/max.cpp [hunk 2]
```diff
@@ -67,6 +68,8 @@ bool op::v1::ReduceMax::evaluate(const HostTensorVector& outputs,
                                  const HostTensorVector& inputs) const
 {
     NGRAPH_OP_SCOPE(v1_ReduceMax_evaluate);
+    NGRAPH_CHECK(validate_host_tensor_vector(inputs, 2));
```
Comment (ilyachur) [inline ngraph/core/src/op/max.cpp:71]: Looks like we don't use the second input, am I right?

Comment (mitruska) [reply ngraph/core/src/op/max.cpp:71]: It should be probably updated to reuse `get_normalized_axes_from_tensor` helper with second input

File: ngraph/core/src/op/max.cpp [hunk 3]
```diff
@@ -67,6 +68,8 @@ bool op::v1::ReduceMax::evaluate(const HostTensorVector& outputs,
...
+    NGRAPH_CHECK(validate_host_tensor_vector(outputs, 1));
     return maxop::evaluate_max(inputs[0], outputs[0], get_reduction_axes(), get_keep_dims());
```
Comment (mitruska) [inline ngraph/core/src/op/max.cpp:73]: Please update to reuse `get_normalized_axes_from_tensor` helper

Comment (ggalieroc) [reply ngraph/core/src/op/max.cpp]: Done

File: ngraph/core/src/op/max.cpp (hunk 2 above)
Comment (ggalieroc) [reply ngraph/core/src/op/max.cpp:72]: I had to wait for ReduceProd to introduce correctly the usage of helper to retrieve axes from host tensor. Now it makes sense to check for second input.

**Review by mitruska — APPROVED** [review]

**Review by tomdol — CHANGES_REQUESTED** [review]
Similarily to the other PR, could you switch from `Backend` to `TestCase` in the tests? If you want the tests to be executed on the interpreter engine only, use exactly this type to parametrize the `TestCase` template.

**Comment by ggalieroc** [comment]
@tomdol I agree that we should switch to `TestCase`, but there are separate tasks created for this in the past, including reduce ops, eg 35880, 35197. Refactor backend tests to TestCase is not part of requirements of ref impl. What we are doing is that every new test introduced uses TestCase. For refactoring, as it involves significant amount of work, we will assign those already created tasks and allocate them in the plan

**Review by tomdol — APPROVED** [review]

**Review by ilyachur — APPROVED** [review]

**Pull request merged**